Add avatars for reactions participants

paulmolluzzo opened the pull request:
Hat tip to @hkdobrev for getting this started and finding the way to get GH user avatars.

This extends the work @hkdobrev did on a new branch that is also up to date with `master`. Basically, it takes the avatars of people who add a reaction to a comment and appends their avatar. I am only appending the first 3 (or less) avatars because it quickly gets crazy, and then just add `+${remainder}` at the end.

Suggestions on the design are welcome. I'm thinking we can even get rid of the original number, but not sure.

Example:
![reaction-avatars](https://cloud.githubusercontent.com/assets/737065/14230374/ca390f4a-f921-11e5-92a8-c1d81ce51b4c.png)

SpaceK33z commented:
Wow, nice work!

DrewML commented on extension/content.js:
```diff
@@ -307,6 +307,43 @@ $(document).on('keydown', event => {
+		const $reactionButtons = $reactionsContainer.find('.comment-reactions-options .reaction-summary-item[aria-label]');
+
+		$reactionButtons.each((index, element) => {
+			const participants = [];
```
Instead of the empty array here, why not just set `const participants` to the return value of the `map` (lines 322-328)

paulmolluzzo replied:
Yup, good idea. Will do.

DrewML commented:
Nice work!

There is one small issue I spotted (that I don't think should preclude the PR, but we should address later). When you "react" to an issue, the user icon doesn't show until a page refresh. At that point, you obviously know which face should be there (your own!), but fixing it would make the UI feel a small bit more consistent.

paulmolluzzo commented:
This closes #95, BTW.

paulmolluzzo commented:
> When you "react" to an issue, the user icon doesn't show until a page refresh.

Can fix. Thanks @DrewML

paulmolluzzo commented:
OK, this is updated as follows:
1. I fixed the `participants` array as suggested by @DrewML to make it easier to read.
2. I changed the `allParticipants` variable name to `participantsCount` because it's more obvious what I'm doing with it.
3.  I added a check to make sure the `div.participants-container` that I use to put avatars in didn't exist. (This is to set up the suggestion to handle the "live" update.)
4. I rebased the whole PR (with one exception). @hkdobrev's original work is still on https://github.com/sindresorhus/refined-github/tree/reactions-participants if we need it anyway.

About the suggestion in https://github.com/sindresorhus/refined-github/pull/145#issuecomment-204932062: I tried to do this, but the only "working" solution I have is to set an event listener with a `window.setTimeout` that does `addReactionParticipants` after 1 second. This is really not ideal, IMO, and kind of hacky. I tried to do this with a mutation observer, but that got super out of control and my browser kept crashing. If anyone has a suggestion please let me know. I didn't rebase that commit because it's still iffy to me.

leocaseiro commented:
Loved it!

How about adding `border-radius: 3px;`?

leocaseiro commented:
I agree you should remove the original number same way facebook does...

sindresorhus commented on extension/content.css:
```diff
@@ -328,3 +328,18 @@ svg.octicon.octicon-star.dashboard-event-icon {
+	margin-left: 3px;
+	vertical-align: middle;
+}
+.reaction-summary-item div img {
```
This is unclosed.

sindresorhus commented on extension/content.js:
```diff
@@ -307,6 +307,48 @@ $(document).on('keydown', event => {
 	}
 });
 
+function addReactionParticipants() {
```
Maybe move this to a separate file?

sindresorhus commented:
> How about adding border-radius: 3px;?

:+1:

sindresorhus commented:
I don't think it should show your own avatar. That's pretty moot and you can already see you've voted.

sindresorhus commented:
> I agree you should remove the original number same way facebook does...

:+1:

sindresorhus commented:
Seems to break the UI when there are a lot of different reactions: https://github.com/facebook/react/issues/6322

paulmolluzzo commented:
> I don't think it should show your own avatar. That's pretty moot and you can already see you've voted.

What if you're the only vote? It's going to look like this, which is weird:
> :+1: +1

> Seems to break the UI when there are a lot of different reactions: facebook/react#6322

I can't see what's broken, can you screenshot?

sindresorhus commented:
> What if you're the only vote? It's going to look like this, which is weird:

Drop the `+` in that case.

> I can't see what's broken, can you screenshot?

<img width="728" alt="screen shot 2016-04-04 at 20 41 48" src="https://cloud.githubusercontent.com/assets/170270/14249793/af4f615e-faa5-11e5-8469-65ead1169d29.png">

Notice the white overflow below the reactions.

hkdobrev commented:
I'd recommend to exclude your own avatar from the list. You already have a different background colour if you have reacted with a certain emoji. No need to listen when you click and update it. Just never include your own avatar in the list.

paulmolluzzo commented:
I pushed changes and didn't squash (we can do that later).

I did everything suggested, but can't replicate that but you guys are seeing with the white space. I might have inadvertently fixed it when I removed the original count? Or maybe it was that unclosed CSS bracket?

paulmolluzzo commented:
> Notice the white overflow below the reactions.

@sindresorhus I can't replicate. Can you try to figure what's happening? Please make sure you're up to date with what I pushed. It might be that the two reaction counts were doing it.

sindresorhus commented:
@paulmolluzzo Yes, I was using an older version of your PR. But even with latest, if you have enough reactions it will break. Might be an edge-case though. I tried adding two more reactions to https://github.com/facebook/react/issues/6322 and this is what I got (using the latest version of this PR):


<img width="734" alt="screen shot 2016-04-04 at 21 26 57" src="https://cloud.githubusercontent.com/assets/170270/14253047/297a90c6-fab4-11e5-96ec-7cb352f521ca.png">

-

Can you fix the merge conflict and add this feature to the readme?

paulmolluzzo commented:
@sindresorhus I replicated it. I tightened up the padding on those buttons to compensate, but this will probably break if the post is super popular. An edge case for sure.

paulmolluzzo commented:
@sindresorhus BTW, moving this to a separate file broke the click event listening. I can't use the `setTimeout` from the new file. 😞 

LMK if you have an idea for a solution or just want to take out the event binding or thing shoving this back into the main content.js is cool.

hkdobrev commented:
@paulmolluzzo I've suggested not listening to click and not showing your own reactions ever.

paulmolluzzo commented:
> I've suggested not listening to click and not showing your own reactions ever.

@hkdobrev if everyone else agrees taking it out is super easy. I'm already not showing your own avatar, but your reaction is part of the total count.

paulmolluzzo commented:
OK, I am pretty confident in the current version, and this time I did rebase everything to get it ready to merge. I'm still open to feedback on everything, so please let me know if anyone finds more bugs.

@hkdobrev, when you don't have the event listener for `click` on the buttons all the avatars just disappear and it _looks_ like a bug, so I do think it should stay. However, I'm not showing the current User's avatar ever and it looks good. If you want to see what happens without the `click`, pull down the PR and comment out this line: https://github.com/sindresorhus/refined-github/pull/145/commits/a0f5ca628be8f1bde9f821f7916bf788a98fa547#diff-8f297dee0ce8264d927a877e48fec8acR12

@sindresorhus This is still in a separate file, and I should have fixed the UI issue correctly. The caveat to that is with enough reactions on enough _different_ reactions it'll eventually break I think but that's probably a super edge case.

// @DrewML

DrewML commented:
This is awesome! Great work @paulmolluzzo!

One issue I'm seeing locally:

When "unreacting" to a reaction that has >= 1 other users that chose the same reaction, clicking on the reaction icon itself (rather than the outer blue area) throwns an exceptions, and doesn't unreact.
```
addreactionparticipants.js:12 Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src assets-cdn.github.com".

element.addEventListener @ addreactionparticipants.js:12
```

I'm getting a similar exception (using the [https://github.com/facebook/react/issues/6322](facebook issue thread) as a test-case) when I try to react using a reaction others have used, but it appears to be a bit sporadic right now.

paulmolluzzo commented:
@DrewML Please pull down the latest change (another rebase, sorry). Chrome won't allow `setTimeout` so I switched to `setInterval`. Go figure.

Also, I took out the anchor tag surrounding the avatars because clicking the image directed you to the user and made it hard to just add a reaction.

DrewML commented on extension/addreactionparticipants.js:
```diff
@@ -0,0 +1,60 @@
+		const $reactionButtons = $reactionsContainer.find('.comment-reactions-options .reaction-summary-item[aria-label]');
+
+		$reactionButtons.each((index, element) => {
+			element.addEventListener('click', () => {
```
Would it be difficult to switch this out for a delegated listener? That's going to be quite a large number of event listeners on long pages with a lot of activity on them.

paulmolluzzo replied:
You mean one listener and checking if it matches the target class? It's not tough, but does it have a big benefit? I don't mind doing it, just curious.

DrewML replied:
It's definitely not a *big* benefit at all - force of habit :stuck_out_tongue:

paulmolluzzo replied:
OK, this is added in. Now there's one event listener and it's only applied once. Before it was reapplying the event listeners every time you clicked.

DrewML commented on extension/addreactionparticipants.js:
```diff
@@ -0,0 +1,60 @@
+const addReactionParticipants = currentUser => {
+	$('.comment-reactions').each((index, reactionsContainer) => {
```
What do you think about having `addReactionParticipants` accept an optional comment container/id, or something along those lines? Right now, it looks like, when a user reacts (line 13), we're going to iterate over and re-render every reaction bar for each comment on the page. Knowing which comment is the target would eliminate that.

paulmolluzzo replied:
Hmm, I can't remember if the socket refreshes the comment's reactions or all reactions. I can take a look.

paulmolluzzo replied:
This would require refactoring the initial selection of the container, but I don't see a benefit unless we pass in the new selection on the click. I don't know that we'll see a big improvement, and there's a positive side effect that if any of the _other_ comments are out of date (because of a socket) they'll get refreshed. I would do this in tandem with the socket thing.

DrewML commented:
Another issue I don't think is a big deal, but I'll point it out just to have it documented.

GitHub pushes updates to the page (websocket) whenever another user has reacted in a comment thread you're viewing. When that happens, they swap out the DOM and our feature goes away for that particular comment.

There are only 2 ways I can think of to identify when updates have been pushed, and I'm not sure that they're worth the effort:

1. Mutation Observer
2. Injecting a script into the page from the content script (to get shared JS execution context), then listen for the `socket:message` event on the document

hkdobrev commented:
@paulmolluzzo Great work! Sorry for being late to the show. Thanks for the nice explanation earlier, I haven't had the time to look into it properly before.

Regarding keeping me as author of the first commit, I'm fine with removing me and squashing however you see fit, most of the code there is yours anyway 😄  Hypothetically if you want to keep one's work I'll prefer the original commit patch and then your commit so it's visible how you've changed it, but as far as I know @sindresorhus prefers squashing more.

The one thing I really don't like about this change is the `+${remainder}` instead of showing the total count like the original GitHub implementation. The "A B C +X more` UX pattern makes you do cognitive work of counting and adding things up. When having multiple type of reactions with low number of reactions per emoji it's harder to compare which one is more popular when you have an emoji with a count and an emoji without a count. In that regard I think the original GitHub implementation is superior. I suggest to still have this great addition to the extension, but closer to the way canonical reactions work and keep the total count for each emoji to the right of the emoji (before the avatars). 

This would also solve one other small thing for me. The emoji and the first avatar are too close to each other and with very similar sizes. Some avatars may confuse people if they look similar to an emoji or even if they are predominantly filled with that distinctive yellow emoji colour. So having the number of reactions between the emoji and the avatars would separate them a bit.

hkdobrev commented on extension/addreactionparticipants.js:
```diff
@@ -0,0 +1,65 @@
+			const $reactionsContainer = $(reactionsContainer);
+			if (!$reactionsContainer.hasClass('has-reactions')) {
+				return;
+			}
```
I think you could move this check to the `$('.comment-reactions')` selector.

sindresorhus commented:
After using this for a while, I have to agree with @hkdobrev.

hkdobrev commented:
> After using this for a while, I have to agree with @hkdobrev.

When I first saw this and the screenshot I was really happy with how much better this is than my initial suggestions in https://github.com/sindresorhus/refined-github/issues/95#issuecomment-199885275 Having up to three avatars next to each reaction is a very nice visual way to see who reacted and if there the same avatars reacted on multiple reactions.

When I started using it though I wanted to revert to GitHub original implementation every time I was using reactions, only because of the total count.

paulmolluzzo commented:
So keep the original count and remove the `remainder`? @hkdobrev @sindresorhus

hkdobrev commented:
@paulmolluzzo 
> So keep the original count and remove the remainder?

That would be my preference, yes.

paulmolluzzo commented:
@hkdobrev @sindresorhus I've changed it so that the original count is being shown, not the `remainder`. I've also moved the test for `.has-reactions` up as @hkdobrev suggested, and I reduced the interval to `500ms` for the click event handler.

sindresorhus commented on readme.md:
```diff
@@ -21,6 +21,8 @@ Our hope is that GitHub will notice and implement some of these much needed impr
 - [Differentiates merge commits from regular commits](https://cloud.githubusercontent.com/assets/170270/14101222/2fe2c24a-f5bd-11e5-8b1f-4e589917d4c4.png)
 - [Hides :+1: :-1: comments and shows their count in the sidebar](https://cloud.githubusercontent.com/assets/170270/13241396/0b708ae8-da1d-11e5-8c01-94eae501034c.png)
+- [Adds User avatars to Reactions](https://cloud.githubusercontent.com/assets/737065/14269443/c80ae418-fab2-11e5-93f7-d776f40a5b25.png)
+- Supports indenting with the tab key in textareas like the comment box
```
This is duplicated. Probably some merge conflict mistake.

sindresorhus commented on readme.md:
```diff
@@ -21,6 +21,8 @@ Our hope is that GitHub will notice and implement some of these much needed impr
 - [Adds links to patch and diff for each commit](https://cloud.githubusercontent.com/assets/737065/13605562/22faa79e-e516-11e5-80db-2da6aa7965ac.png)
 - [Differentiates merge commits from regular commits](https://cloud.githubusercontent.com/assets/170270/14101222/2fe2c24a-f5bd-11e5-8b1f-4e589917d4c4.png)
 - [Hides :+1: :-1: comments and shows their count in the sidebar](https://cloud.githubusercontent.com/assets/170270/13241396/0b708ae8-da1d-11e5-8c01-94eae501034c.png)
+- [Adds User avatars to Reactions](https://cloud.githubusercontent.com/assets/737065/14269443/c80ae418-fab2-11e5-93f7-d776f40a5b25.png)
```
lowercase `User`

sindresorhus commented on extension/manifest.json:
```diff
@@ -29,6 +29,7 @@
 				"vendor/gh-injection.js",
 				"page-detect.js",
 				"diffheader.js",
+				"addreactionparticipants.js",
```
`reactions-avatars.js`?

sindresorhus commented on extension/content.css:
```diff
@@ -339,3 +339,29 @@ svg.octicon.octicon-star.dashboard-event-icon {
 	height: 200px !important;
 }
+
+/* styles for reaction avatars */
```
Specify that it's for our implementation:

> styles for our reactions avatar feature

hkdobrev commented on extension/content.js:
```diff
@@ -372,6 +372,10 @@ document.addEventListener('DOMContentLoaded', () => {
+				addReactionParticipants.add(username);
+				document.addEventListener('click', event => {
+					addReactionParticipants.reapply(event, username);
+				});
```
Reactions are also present on comments of commits. E.g. https://github.com/mathiasbynens/dotfiles/commit/fb0bd713f62d1665f4cabed1aa97f7763f18f0b3#commitcomment-16984320

paulmolluzzo replied:
@hkdobrev fixed. Thanks.

paulmolluzzo commented:
@sindresorhus fixed all the issues you pointed out and got this up to date with `master`.

hkdobrev commented on extension/reactions-avatars.js:
```diff
@@ -0,0 +1,58 @@
+		document.addEventListener('click', event => {
+			addReactionParticipants.reapply(event, username);
+		});
+	}
```
You could rename `username` to `currentUser` for consistency and even put the `reapply` function body in `addListener` itself as reapplying only makes sense at the moment in the context of an event listener.

paulmolluzzo replied:
Changed the `username` param.

I'm not going to move the `reapply` function because leaving it here allows us to use it separately in another event listener or whatever.

hkdobrev commented:
I've been using the updated version and I really love it! ❤️

hkdobrev commented on extension/content.js:
```diff
@@ -390,6 +392,8 @@ document.addEventListener('DOMContentLoaded', () => {
 				addPatchDiffLinks();
+				addReactionParticipants.add(username);
+				addReactionParticipants.addListener(username);
 			}
```
You could combine these in a single `if (pageDetect.isPR() || pageDetect.isIssue() || pageDetect.isCommit())`. I think it's more consistent, but it would also help with the fact `isCommit()` would return `true` for PR commits which match `isPR()` as well. So in some cases you would match both ifs and execute this twice. With a single if with 3 or expressions, you would execute it only once.

paulmolluzzo replied:
Pushed.

hkdobrev commented:
LGTM 💫  🎉  🎊

hkdobrev commented on readme.md:
```diff
@@ -21,6 +21,8 @@ Our hope is that GitHub will notice and implement some of these much needed impr
 - [Adds links to patch and diff for each commit](https://cloud.githubusercontent.com/assets/737065/13605562/22faa79e-e516-11e5-80db-2da6aa7965ac.png)
 - [Differentiates merge commits from regular commits](https://cloud.githubusercontent.com/assets/170270/14101222/2fe2c24a-f5bd-11e5-8b1f-4e589917d4c4.png)
 - [Hides :+1: :-1: comments and shows their count in the sidebar](https://cloud.githubusercontent.com/assets/170270/13241396/0b708ae8-da1d-11e5-8c01-94eae501034c.png)
+- [Adds user avatars to Reactions](https://cloud.githubusercontent.com/assets/737065/14269443/c80ae418-fab2-11e5-93f7-d776f40a5b25.png)
```
Could you update the screenshot contents?

paulmolluzzo commented:
Updated the README @hkdobrev

sindresorhus commented:
Yay! 🎉✨ This is super good @paulmolluzzo. Thanks for working so hard on it :)

![687474703a2f2f7777772e7265616374696f6e676966732e636f6d2f77702d636f6e74656e742f67616c6c6572792f64616e63652d70617274792f74756d626c725f6d306f73727268714d6b31717a6376376e6f345f3235302e676966](https://cloud.githubusercontent.com/assets/170270/14328953/5828fb6e-fc63-11e5-9425-ec6224903598.gif)

DrewML commented:
@paulmolluzzo Awesome work (and of course thanks @hkdobrev for getting this one started).
![P2MB29d7V3rhu](https://camo.githubusercontent.com/af8b39e68cb3cb730458c1921fdf42abfb108be0/68747470733a2f2f6d65646961342e67697068792e636f6d2f6d656469612f50324d423239643756337268752f67697068792e676966)

hkdobrev commented:
@paulmolluzzo Thanks for your very hard work on this! I know a code review so big and all of the amends are hard, but you've plowed through it flawlessly! The feature is working really great!

hkdobrev commented:
@sindresorhus the squashed commit you've made to master is attributed to me though which is not really fair.

paulmolluzzo commented:
Thanks everyone! Would like to see an event listener for the socket if @DrewML has an idea there.

Thanks @hkdobrev for getting the ball rolling and finding that secret avatar URL, this pretty much only works because of that.

hkdobrev commented:
@sindresorhus Do you think we could tag and land this in the web store?

sindresorhus commented:
@hkdobrev Done
